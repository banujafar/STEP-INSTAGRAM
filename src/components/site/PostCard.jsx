import React from "react";
import { Link } from "react-router-dom";
import { AiFillHeart } from "react-icons/ai";
import { FaComment } from "react-icons/fa";

const PostCard = ({ post }) => {
  return (
    <Link to="#" className="w-64 h-60 mr-6 mt-12">
      <div className="w-full h-full relative group/item shadow-lg cursor-pointer">
        <img src={post.imageUrl} className="w-full h-full" alt="Post" />
        <div className="bg-opacity-0 hover:bg-black-rgba transition-all duration-500 w-full h-full absolute left-0 top-0">
          <span className="group/edit invisible group-hover/item:visible absolute left-1/3 top-1/2 flex">
            <LikeButton post={post} />
            <CommentButton post={post} />
          </span>
        </div>
      </div>
    </Link>
  );
};

const LikeButton = ({ post }) => {
  return (
    <span className="flex justify-center text-xl items-center mr-6 group-hover/item:text-white">
      <AiFillHeart className="mr-1" fill="white" size={"25px"} />
      {post.likes.length}
    </span>
  );
};

const CommentButton = ({ post }) => {
  return (
    <span className="flex justify-center text-xl items-center mr-6 group-hover/item:text-white">
      <FaComment fill="white" className="mr-1" size={"25px"} />
      {post.comments.length}
    </span>
  );
};
export default PostCard;
