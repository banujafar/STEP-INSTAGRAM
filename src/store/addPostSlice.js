import { createSlice } from "@reduxjs/toolkit";

const initialState = {
    base64Image: null,
    step: 1,

}

const addPostSlice = createSlice({
    name: 'addPost',
    initialState,
    reducers: {
        changeStep: (state, action) => {
            state.step = action.payload === 'back' ? state.step - 1 : state.step + 1
        },
        
        clearData: (state) => {
            state.base64Image = null
            state.step = 1
        },
        setBase64Image: (state, action) => {
            
            state.base64Image = action.payload
        }


    }
})

export default addPostSlice.reducer
export const { changeStep, setBase64Image, clearData } = addPostSlice.actions