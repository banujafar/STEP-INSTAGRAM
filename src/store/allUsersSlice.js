// userSlice.js

import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";

const initialState = {
  userData: [],
  loading: false,
  error: null,
};

const token = window.localStorage.getItem('access-token') || null;

export const fetchAllUsers = createAsyncThunk(
  "users/fetchAllUsers",
  async (_, thunkAPI) => {
    try {
      const response = await fetch(
        "https://your-api-endpoint-for-all-users",
        {
          method: "GET",
          headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${token}`,
          },
        }
      );

      if (!response.ok) {
        throw new Error("Failed to fetch all users");
      }

      const data = await response.json();
      return data.data;
    } catch (error) {
      return thunkAPI.rejectWithValue(error.message);
    }
  }
);

const allUsersSlice = createSlice({
  name: "users",
  initialState,
  extraReducers: (builder) => {
    builder.addCase(fetchAllUsers.pending, (state) => {
      state.loading = true;
    });
    builder.addCase(fetchAllUsers.fulfilled, (state, action) => {
      state.loading = false;
      state.userData = action.payload;
      state.error = null;
    });
    builder.addCase(fetchAllUsers.rejected, (state, action) => {
      state.loading = false;
      state.userData = [];
      state.error = action.payload;
    });
  },
});

export const userActions = allUsersSlice.actions;

export default allUsersSlice.reducer;
