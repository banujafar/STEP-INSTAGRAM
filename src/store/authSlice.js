import { createSlice } from "@reduxjs/toolkit";


const initialState = {
    token: window['localStorage'].getItem('access-token') || null,
    username: window['localStorage'].getItem('username') || null,
    userData: {}
}

const authSlice = createSlice({
    name: 'auth',
    initialState,
    reducers: {
        setCredentials: (state, action) => {
            const { token, username } = action.payload
            state.token = token
            state.username = username
            window['localStorage'].setItem('access-token', token)
            window['localStorage'].setItem('username', username)
        },
        logOut: state => {
            state.token = null
            state.username = null
            window['localStorage'].removeItem('access-token')
            window['localStorage'].removeItem('username')
        },
        setCurrentUser: (state, action) => {
            state.userData = action.payload
        }
    },

})

export default authSlice.reducer
export const { setCredentials, logOut, setCurrentUser } = authSlice.actions

export const selectCurrentToken = (state) => state.auth.token