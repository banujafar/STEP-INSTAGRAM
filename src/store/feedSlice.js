// feedSlice.js

import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";

const initialState = {
  feedData: [],
  loading: false,
  error: null,
};



export const fetchUserFeed = createAsyncThunk(
  "feed/fetchUserFeed",
  async (token, thunkAPI) => {
    try {
      const response = await fetch(
        "https://instagram.brightly-shining.cloud/api/v1/user/feed",
        {
          method: "GET",
          headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${token}`,
          },
        }
      );

      if (!response.ok) {
        throw new Error("Failed to fetch user feed");
      }

      const data = await response.json();
      return data.data;
    } catch (error) {
      return thunkAPI.rejectWithValue(error.message);
    }
  }
);


export const followUser = createAsyncThunk(
  "feed/followUser",
  async (username, thunkAPI) => {
    try {
      const response = await fetch(
        `https://instagram.brightly-shining.cloud/api/v1/user/subscription`,
        {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${token}`,
          },
          body: JSON.stringify({ username }),
        }
      );

      if (!response.ok) {
        throw new Error("Failed to follow user");
      }

      return username;
    } catch (error) {
      return thunkAPI.rejectWithValue(error.message);
    }
  }
);

export const unfollowUser = createAsyncThunk(
  "feed/unfollowUser",
  async (username,token, thunkAPI) => {
   try {

      const response = await fetch(
        `https://instagram.brightly-shining.cloud/api/v1/user/subscription?username=${username}`,
        {
          method: "DELETE",
          headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${token}`,
          },
        }
      );

      if (!response.ok) {
        throw new Error("Failed to unfollow user");
      }

      return username;
    } catch (error) {
      return thunkAPI.rejectWithValue(error.message);
    }
  }
);


const feedSlice = createSlice({
  name: "feed",
  initialState,
  extraReducers: (builder) => {
    builder.addCase(fetchUserFeed.pending, (state) => {
      state.loading = true;
    });
    builder.addCase(fetchUserFeed.fulfilled, (state, action) => {
      state.loading = false;
      state.feedData = action.payload;
      state.error = null;
    });
    builder.addCase(fetchUserFeed.rejected, (state, action) => {
      state.loading = false;
      state.feedData = [];
      state.error = action.payload;
    });
  },
});

export const feedActions = feedSlice.actions;

export default feedSlice.reducer;
