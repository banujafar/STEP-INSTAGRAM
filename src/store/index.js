import { configureStore } from "@reduxjs/toolkit";

import authReducer from "./authSlice";
import { apiSlice } from "./api/apiSlice";
import modalReducer from "./modalSlice";
import feedReducer from "./feedSlice";

const store = configureStore({
    reducer: {
        [apiSlice.reducerPath]: apiSlice.reducer,
        auth: authReducer,

        modal: modalReducer,
        feed: feedReducer,
    },
    middleware: getDefaultMiddleware => getDefaultMiddleware().concat(apiSlice.middleware),
    devTools: true
});
export default store;