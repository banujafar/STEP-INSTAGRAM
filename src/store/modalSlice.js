import { createSlice } from "@reduxjs/toolkit";



const initialState = {
    modals: []
}

const modalSlice = createSlice({
    name: 'modal',
    initialState,
    reducers: {
        appendModal: (state, action) => {
            
            state.modals.push(action.payload)
        },
        deleteModal: (state) => {
            state.modals.slice(0, state.modals.length - 1)
        }
    }
})

export default modalSlice.reducer
export const {appendModal, deleteModal} = modalSlice.actions