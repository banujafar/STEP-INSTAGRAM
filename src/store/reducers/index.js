import { combineReducers } from "redux";
import { userListReducer } from "./userList";


const rootReducer = combineReducers({
    userList: userListReducer
})

export default rootReducer