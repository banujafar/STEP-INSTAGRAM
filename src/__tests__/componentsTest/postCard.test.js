import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import PostCard from "../../components/site/PostCard";
import { Provider } from "react-redux";
import store from "../../store";

describe("render post card component", () => {
  const post = {
    authorUsername: "alexbird",
    caption: "Photo was made from scanned 35mm film",
    postId: "02c6f4fdefff4409b12b793cfb5e0eeb",
    timestamp: 1693024855,
    likes: [{ authorUsername: "kate_oliver" }],
  };

  it("when click post open Modal", async () => {
    render(
      <Provider store={store}>
        <PostCard post={post} />
      </Provider>
    );

    const openModalBtn = screen.getByTestId("post-modal");
    fireEvent.click(openModalBtn);

    const expectedAction = {
      modalName: "post modal",
      data: {
        postId: "02c6f4fdefff4409b12b793cfb5e0eeb",
      },
    };

    expect(store.getState().modal.modals[0]).toEqual(expectedAction);
  });

  it("snapshot testing", () => {
    const tree = render(<PostCard post={post} />);
    expect(tree).toMatchSnapshot();
  });
});
