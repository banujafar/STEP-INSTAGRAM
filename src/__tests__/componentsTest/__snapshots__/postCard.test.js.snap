// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostCard Component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-72 h-60 mr-2.5 mt-12"
    data-testid="post-modal"
  >
    <div
      class="w-full h-full relative group/item shadow-lg cursor-pointer"
    >
      <img
        alt="Post"
        class="w-full h-full"
      />
      <div
        class="bg-opacity-0 hover:bg-black-rgba transition-all duration-500 w-full h-full absolute left-0 top-0"
      >
        <span
          class="group/edit invisible group-hover/item:visible absolute left-1/3 top-1/2 flex"
        >
          <div
            class="flex justify-start flex-col"
          >
            <div
              class="flex items-center space-x-4"
            >
              <button
                class="flex items-center text-gray-600"
              >
                <svg
                  class="text-2xl"
                  fill="white"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 1024 1024"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M923 283.6a260.04 260.04 0 0 0-56.9-82.8 264.4 264.4 0 0 0-84-55.5A265.34 265.34 0 0 0 679.7 125c-49.3 0-97.4 13.5-139.2 39-10 6.1-19.5 12.8-28.5 20.1-9-7.3-18.5-14-28.5-20.1-41.8-25.5-89.9-39-139.2-39-35.5 0-69.9 6.8-102.4 20.3-31.4 13-59.7 31.7-84 55.5a258.44 258.44 0 0 0-56.9 82.8c-13.9 32.3-21 66.6-21 101.9 0 33.3 6.8 68 20.3 103.3 11.3 29.5 27.5 60.1 48.2 91 32.8 48.9 77.9 99.9 133.9 151.6 92.8 85.7 184.7 144.9 188.6 147.3l23.7 15.2c10.5 6.7 24 6.7 34.5 0l23.7-15.2c3.9-2.5 95.7-61.6 188.6-147.3 56-51.7 101.1-102.7 133.9-151.6 20.7-30.9 37-61.5 48.2-91 13.5-35.3 20.3-70 20.3-103.3.1-35.3-7-69.6-20.9-101.9z"
                  />
                </svg>
                <span
                  class="text-white ml-2"
                >
                  1
                </span>
              </button>
              <button
                class="flex items-center text-gray-600"
              >
                <svg
                  class="text-2xl"
                  fill="white"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M256 32C114.6 32 0 125.1 0 240c0 49.6 21.4 95 57 130.7C44.5 421.1 2.7 466 2.2 466.5c-2.2 2.3-2.8 5.7-1.5 8.7S4.8 480 8 480c66.3 0 116-31.8 140.6-51.4 32.7 12.3 69 19.4 107.4 19.4 141.4 0 256-93.1 256-208S397.4 32 256 32z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`render post card component snapshot testing 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="w-72 h-60 mr-2.5 mt-12"
        data-testid="post-modal"
      >
        <div
          class="w-full h-full relative group/item shadow-lg cursor-pointer"
        >
          <img
            alt="Post"
            class="w-full h-full"
          />
          <div
            class="bg-opacity-0 hover:bg-black-rgba transition-all duration-500 w-full h-full absolute left-0 top-0"
          >
            <span
              class="group/edit invisible group-hover/item:visible absolute left-1/3 top-1/2 flex"
            >
              <div
                class="flex justify-start flex-col"
              >
                <div
                  class="flex items-center space-x-4"
                >
                  <button
                    class="flex items-center text-gray-600"
                  >
                    <svg
                      class="text-2xl"
                      fill="white"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 1024 1024"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M923 283.6a260.04 260.04 0 0 0-56.9-82.8 264.4 264.4 0 0 0-84-55.5A265.34 265.34 0 0 0 679.7 125c-49.3 0-97.4 13.5-139.2 39-10 6.1-19.5 12.8-28.5 20.1-9-7.3-18.5-14-28.5-20.1-41.8-25.5-89.9-39-139.2-39-35.5 0-69.9 6.8-102.4 20.3-31.4 13-59.7 31.7-84 55.5a258.44 258.44 0 0 0-56.9 82.8c-13.9 32.3-21 66.6-21 101.9 0 33.3 6.8 68 20.3 103.3 11.3 29.5 27.5 60.1 48.2 91 32.8 48.9 77.9 99.9 133.9 151.6 92.8 85.7 184.7 144.9 188.6 147.3l23.7 15.2c10.5 6.7 24 6.7 34.5 0l23.7-15.2c3.9-2.5 95.7-61.6 188.6-147.3 56-51.7 101.1-102.7 133.9-151.6 20.7-30.9 37-61.5 48.2-91 13.5-35.3 20.3-70 20.3-103.3.1-35.3-7-69.6-20.9-101.9z"
                      />
                    </svg>
                    <span
                      class="text-white ml-2"
                    >
                      1
                    </span>
                  </button>
                  <button
                    class="flex items-center text-gray-600"
                  >
                    <svg
                      class="text-2xl"
                      fill="white"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M256 32C114.6 32 0 125.1 0 240c0 49.6 21.4 95 57 130.7C44.5 421.1 2.7 466 2.2 466.5c-2.2 2.3-2.8 5.7-1.5 8.7S4.8 480 8 480c66.3 0 116-31.8 140.6-51.4 32.7 12.3 69 19.4 107.4 19.4 141.4 0 256-93.1 256-208S397.4 32 256 32z"
                      />
                    </svg>
                  </button>
                </div>
              </div>
            </span>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="w-72 h-60 mr-2.5 mt-12"
      data-testid="post-modal"
    >
      <div
        class="w-full h-full relative group/item shadow-lg cursor-pointer"
      >
        <img
          alt="Post"
          class="w-full h-full"
        />
        <div
          class="bg-opacity-0 hover:bg-black-rgba transition-all duration-500 w-full h-full absolute left-0 top-0"
        >
          <span
            class="group/edit invisible group-hover/item:visible absolute left-1/3 top-1/2 flex"
          >
            <div
              class="flex justify-start flex-col"
            >
              <div
                class="flex items-center space-x-4"
              >
                <button
                  class="flex items-center text-gray-600"
                >
                  <svg
                    class="text-2xl"
                    fill="white"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 1024 1024"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M923 283.6a260.04 260.04 0 0 0-56.9-82.8 264.4 264.4 0 0 0-84-55.5A265.34 265.34 0 0 0 679.7 125c-49.3 0-97.4 13.5-139.2 39-10 6.1-19.5 12.8-28.5 20.1-9-7.3-18.5-14-28.5-20.1-41.8-25.5-89.9-39-139.2-39-35.5 0-69.9 6.8-102.4 20.3-31.4 13-59.7 31.7-84 55.5a258.44 258.44 0 0 0-56.9 82.8c-13.9 32.3-21 66.6-21 101.9 0 33.3 6.8 68 20.3 103.3 11.3 29.5 27.5 60.1 48.2 91 32.8 48.9 77.9 99.9 133.9 151.6 92.8 85.7 184.7 144.9 188.6 147.3l23.7 15.2c10.5 6.7 24 6.7 34.5 0l23.7-15.2c3.9-2.5 95.7-61.6 188.6-147.3 56-51.7 101.1-102.7 133.9-151.6 20.7-30.9 37-61.5 48.2-91 13.5-35.3 20.3-70 20.3-103.3.1-35.3-7-69.6-20.9-101.9z"
                    />
                  </svg>
                  <span
                    class="text-white ml-2"
                  >
                    1
                  </span>
                </button>
                <button
                  class="flex items-center text-gray-600"
                >
                  <svg
                    class="text-2xl"
                    fill="white"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 512 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M256 32C114.6 32 0 125.1 0 240c0 49.6 21.4 95 57 130.7C44.5 421.1 2.7 466 2.2 466.5c-2.2 2.3-2.8 5.7-1.5 8.7S4.8 480 8 480c66.3 0 116-31.8 140.6-51.4 32.7 12.3 69 19.4 107.4 19.4 141.4 0 256-93.1 256-208S397.4 32 256 32z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </span>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
